llvm/lib/Transforms/InstCombine/InstCombineAndOrXor.cpp

```diff
@@ -591,10 +592,9 @@ static Value *foldLogOpOfMaskedICmps(ICmpInst *LHS, ICmpInst *RHS, bool IsAnd,
   // their actual values. This isn't strictly necessary, just a "handle the
   // easy cases for now" decision.
   const APInt *ConstB, *ConstD;
-  if (!match(B, m_APInt(ConstB)) || !match(D, m_APInt(ConstD)))
-    return nullptr;
+  bool DBconst = match(B, m_APInt(ConstB)) && match(D, m_APInt(ConstD));
 
-  if (Mask & (Mask_NotAllZeros | BMask_NotAllOnes)) {
+  if (DBconst && Mask & (Mask_NotAllZeros | BMask_NotAllOnes)) {
```

Comment:
might be cleaner to just nest all of the stuff that relies on B/D being const in an if.